h5py/h5d.pyx

```diff
@@ -534,47 +539,59 @@ cdef class DatasetID(ObjectID):
             compute the raw data. So the default value of `0` means that all
             defined filters have been applied to the raw data.
 
+            If the `out` argument is not None, it must be a writeable
+            contiguous 1D array-like of bytes (e.g., `bytearray` or
+            `numpy.ndarray`) and large enough to contain the whole chunk.
+
             Feature requires: 1.10.2 HDF5
             """
-
             cdef hid_t dset_id
             cdef hid_t dxpl_id
-            cdef hid_t space_id = 0
+            cdef hid_t space_id
             cdef hsize_t *offset = NULL
-            cdef size_t data_size
             cdef int rank
             cdef uint32_t filters
-            cdef hsize_t read_chunk_nbytes
-            cdef char *data = NULL
-            cdef bytes ret
+            cdef hsize_t chunk_bytes, out_bytes
+            cdef int nb_offsets = len(offsets)
+            cdef void * chunk_buffer
 
             dset_id = self.id
             dxpl_id = pdefault(dxpl)
-            space_id = H5Dget_space(self.id)
+            space_id = H5Dget_space(dset_id)
             rank = H5Sget_simple_extent_ndims(space_id)
+            H5Sclose(space_id)
 
-            if len(offsets) != rank:
-                raise TypeError("offset length (%d) must match dataset rank (%d)" % (len(offsets), rank))
+            if nb_offsets != rank:
+                raise TypeError(
+                    f"offsets length ({nb_offsets}) must match dataset rank ({rank})"
+                )
 
+            offset = <hsize_t*>emalloc(sizeof(hsize_t)*rank)
             try:
-                offset = <hsize_t*>emalloc(sizeof(hsize_t)*rank)
                 convert_tuple(offsets, offset, rank)
-                H5Dget_chunk_storage_size(dset_id, offset, &read_chunk_nbytes)
-                data = <char *>emalloc(read_chunk_nbytes)
+                H5Dget_chunk_storage_size(dset_id, offset, &chunk_bytes)
+
+                if out is None:
+                    retval = PyBytes_FromStringAndSize(NULL, chunk_bytes)
+                    chunk_buffer = PyBytes_AsString(retval)
+                else:
+                    out_bytes = out.shape[0]  # Fast way to get out length
+                    if out_bytes < chunk_bytes:
+                        raise ValueError(
+                            f"out buffer is only {out_bytes} bytes, {chunk_bytes} bytes required"
+                        )
+                    retval = memoryview(out[:chunk_bytes])
+                    chunk_buffer = &out[0]
```

Comment:
Just to check, what happens if you try to read an unallocated chunk (which has storage size 0) with a zero-sized buffer? I saw that you turned off bounds checking, so I just want to be sure it won't segfault.

Reply:
It fails in HDF5's read chunk function, the same way as without the `out` argument:

```python
f = h5py.File("test.h5", "w")
ds = f.create_dataset("data", shape=(100,), dtype=numpy.uint16, chunks=(10,), compression='gzip')
f.flush()

out = bytearray(ds.dtype.itemsize * numpy.prod(ds.chunks))
ds.id.read_direct_chunk((0,), out=out)
```
```
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
Input In [42], in <cell line: 1>()
----> 1 ds.id.read_direct_chunk((0,), out=out)

File h5py/h5d.pyx:587, in h5py.h5d.DatasetID.read_direct_chunk()
    585 
    586                 IF HDF5_VERSION >= (1, 10, 3):
--> 587                     H5Dread_chunk(dset_id, dxpl_id, offset, &filters, chunk_buffer)
    588                 ELSE:
    589                     H5DOread_chunk(dset_id, dxpl_id, offset, &filters, chunk_buffer)

OSError: Can't read unprocessed chunk data (storage is not initialized)
```

Worth a test?

Reply:
No, I think that's OK without a test. I don't want to check that it raises OSError because I don't think OSError is quite right, but I don't know what is right, so that's good enough for now. I think there are various places where we misuse OSError (and possibly other exception types).